sw/host/opentitantool/src/command/image.rs

```diff
@@ -160,77 +168,54 @@ impl CommandDispatch for ManifestUpdateCommand {
-            buf.bytes[0..key_bytes.len()].copy_from_slice(key_bytes);
-            image.set_extension_entry(0, MANIFEST_EXT_ID_SPX_KEY, image.size as u32)?;
-            image.add_extension_data(
-                ManifestExtSpxKey {
-                    header: ManifestExtHeader {
-                        identifier: MANIFEST_EXT_ID_SPX_KEY,
-                        name: MANIFEST_EXT_NAME_SPX_KEY,
-                    },
-                    key: *LayoutVerified::<_, SigverifySpxKey>::new(&buf.bytes[0..key_bytes.len()])
-                        .unwrap()
-                        .into_ref(),
-                }
-                .as_bytes(),
-            )?;
+            let key_ext = ManifestExtEntry::new_spx_key_entry(&key)?;
+            image.add_manifest_extension(key_ext)?;
+
             if let SpxKey::Private(private) = key {
                 spx_private_key = Some(private);
             }
-
-            // Update `signed_region_end` after adding all signed extensions (which is just SPX key
-            // extension for now).
-            image.update_signed_region_end()?;
         }
 
-        if self.spx_key.is_some() {
-            image.set_extension_entry(1, MANIFEST_EXT_ID_SPX_SIGNATURE, image.size as u32)?;
-            image.add_extension_data(
-                ManifestExtSpxSignature {
-                    header: ManifestExtHeader {
-                        identifier: MANIFEST_EXT_ID_SPX_SIGNATURE,
-                        name: MANIFEST_EXT_NAME_SPX_SIGNATURE,
-                    },
-                    ..Default::default()
-                }
-                .as_bytes(),
+        // Determine if an SPX+ signature will be added and preallocate the extension.
+        if self.spx_signature.is_some() || spx_private_key.is_some() {
+            ext.check_for_duplicates(ManifestExtId::spx_signature)?;
+            image.allocate_manifest_extension(
+                ManifestExtId::spx_signature.into(),
+                std::mem::size_of::<ManifestExtSpxSignature>(),
             )?;
         }
 
         // Update `length` to be able to generate signatures.
-        image.update_length()?;
-
-        // Apply the manifest values to the image.
-        if let Some(manifest) = &self.manifest {
-            let def = ManifestSpec::read_from_file(manifest)?;
-            image.overwrite_manifest(def)?;
+        if update_length {
+            image.update_length()?;
         }
 
+        // List out all signed extensions and set the bounds of the signed region.
+        // The `spx_key` extension may appear in this list twice, but that doesn't change the
+        // semantics of `update_signed_region()`.
+        let signed_ids = ext
+            .signed_region
+            .iter()
+            .map(|e| e.id())
+            .chain(vec![ManifestExtId::spx_key.into()])
+            .collect::<Vec<u32>>();
+        image.update_signed_region(&signed_ids)?;
+
+        image.drop_null_extensions()?;
```

Comment:
nit: can you clarify where these null extensions come from?

Reply:
maybe dropping a comment explaining that ^ might be good here